### ISSUE TITLE
Latest Salt will not install on Amazon Linux due to missing python, systemd packages and/or EPEL package out of date

### ISSUE THREAD

**Issue by edhgoose** [saltstack/salt#30227]
Hi,

I think this could be interpreted as a documentation bug or an installation bug, but the default instructions for installing Salt on Amazon Linux do not work. I'm attempting to install on an existing AWS Linux image, and I've also reproduced on a brand new Amazon t2.micro instance.

The [installation instructions](https://docs.saltstack.com/en/latest/topics/installation/rhel.html) say that: "Salt should work properly with all mainstream derivatives of Red Hat Enterprise Linux, including CentOS, Scientific Linux, Oracle Linux, and _Amazon Linux_."

However, if I follow the instructions there the installation fails due to python-cardet being out of date and systemd-units, systemd-python missing (full log is further below if that helps):

````
Error: Package: python-requests-2.6.0-1.el7.noarch (saltstack-repo)
           Requires: python-chardet >= 2.2.1-1
           Available: python26-chardet-2.0.1-7.7.amzn1.noarch (amzn-main)
               python-chardet = 2.0.1-7.7.amzn1
Error: Package: salt-2015.8.3-1.el7.noarch (saltstack-repo)
           Requires: systemd-units
Error: Package: salt-master-2015.8.3-1.el7.noarch (saltstack-repo)
           Requires: systemd-python
Error: Package: python-requests-2.6.0-1.el7.noarch (saltstack-repo)
           Requires: python-urllib3 >= 1.10.2-1
           Available: python26-urllib3-1.8.2-1.5.amzn1.noarch (amzn-main)
               python-urllib3 = 1.8.2-1.5.amzn1
Error: Package: salt-2015.8.3-1.el7.noarch (saltstack-repo)
           Requires: systemd-python
````

From looking around (I'm out of my depth here), it looks like systemd is not supported by Amazon Linux (example comment [here](https://forums.aws.amazon.com/message.jspa?messageID=539557)). I'm not sure about the other packages.

I have also tried to use the Epel installation instructions (which work fine, and I've been using them successfully for a while) but the latest version is only 2015.5.5 and I'm trying to confirm a bug is resolved that I previously raised which is only in 2015.8.2 (#28203).

Is there something I'm missing around using systemd? Are the instructions out of date? Is EPEL likely to be updated (who manages that... I couldn't find an answer?).

Please let me know how I can help with testing/fixing this process too - I'd be happy to.

Kind Regards,
Ed

Full Log:

````
$ rpm --import https://repo.saltstack.com/yum/redhat/7/x86_64/latest/SALTSTACK-GPG-KEY.pub

$ vim /etc/yum.repos.d/saltstack.repo

$ sudo yum clean expire-cache
Loaded plugins: priorities, update-motd, upgrade-helper
Cleaning repos: amzn-main amzn-updates saltstack-repo
4 metadata files removed

$ yum install salt-master
Loaded plugins: priorities, update-motd, upgrade-helper
Resolving Dependencies
--> Running transaction check
---> Package salt-master.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: salt = 2015.8.3-1.el7 for package: salt-master-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-python for package: salt-master-2015.8.3-1.el7.noarch
--> Running transaction check
---> Package salt.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: python-msgpack > 0.3 for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: python-tornado >= 4.2.1 for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: python-requests >= 1.0.0 for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: python-futures >= 2.0 for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: python-crypto >= 2.6.1 for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-units for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-units for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-python for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: python-zmq for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: python-six for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: python-markupsafe for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: python-jinja2 for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: PyYAML for package: salt-2015.8.3-1.el7.noarch
---> Package salt-master.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: systemd-python for package: salt-master-2015.8.3-1.el7.noarch
--> Running transaction check
---> Package PyYAML.x86_64 0:3.11-1.el7 will be installed
---> Package python-futures.noarch 0:3.0.3-1.el7 will be installed
---> Package python-msgpack.x86_64 0:0.4.6-1.el7 will be installed
---> Package python-requests.noarch 0:2.6.0-1.el7 will be installed
--> Processing Dependency: python-urllib3 >= 1.10.2-1 for package: python-requests-2.6.0-1.el7.noarch
--> Processing Dependency: python-chardet >= 2.2.1-1 for package: python-requests-2.6.0-1.el7.noarch
---> Package python-tornado.x86_64 0:4.2.1-1.el7 will be installed
--> Processing Dependency: python-pycurl for package: python-tornado-4.2.1-1.el7.x86_64
--> Processing Dependency: python-backports-ssl_match_hostname for package: python-tornado-4.2.1-1.el7.x86_64
---> Package python-zmq.x86_64 0:14.7.0-1.el7 will be installed
--> Processing Dependency: libzmq.so.4()(64bit) for package: python-zmq-14.7.0-1.el7.x86_64
---> Package python26-crypto.x86_64 0:2.6.1-1.12.amzn1 will be installed
--> Processing Dependency: python(abi) = 2.6 for package: python26-crypto-2.6.1-1.12.amzn1.x86_64
--> Processing Dependency: python26(dist-packages) for package: python26-crypto-2.6.1-1.12.amzn1.x86_64
--> Processing Dependency: libpython2.6.so.1.0()(64bit) for package: python26-crypto-2.6.1-1.12.amzn1.x86_64
---> Package python26-jinja2.noarch 0:2.7.2-2.15.amzn1 will be installed
--> Processing Dependency: python26-babel for package: python26-jinja2-2.7.2-2.15.amzn1.noarch
---> Package python26-markupsafe.x86_64 0:0.11-4.6.amzn1 will be installed
---> Package python26-six.noarch 0:1.8.0-1.23.amzn1 will be installed
---> Package salt.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: systemd-units for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-units for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-python for package: salt-2015.8.3-1.el7.noarch
---> Package salt-master.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: systemd-python for package: salt-master-2015.8.3-1.el7.noarch
--> Running transaction check
---> Package python-requests.noarch 0:2.6.0-1.el7 will be installed
--> Processing Dependency: python-urllib3 >= 1.10.2-1 for package: python-requests-2.6.0-1.el7.noarch
--> Processing Dependency: python-chardet >= 2.2.1-1 for package: python-requests-2.6.0-1.el7.noarch
---> Package python26.x86_64 0:2.6.9-2.84.amzn1 will be installed
---> Package python26-babel.noarch 0:0.9.4-5.1.8.amzn1 will be installed
--> Processing Dependency: python26-setuptools for package: python26-babel-0.9.4-5.1.8.amzn1.noarch
---> Package python26-backports-ssl_match_hostname.noarch 0:3.4.0.2-1.12.amzn1 will be installed
--> Processing Dependency: python26-backports for package: python26-backports-ssl_match_hostname-3.4.0.2-1.12.amzn1.noarch
---> Package python26-libs.x86_64 0:2.6.9-2.84.amzn1 will be installed
---> Package python26-pycurl.x86_64 0:7.19.0-17.12.amzn1 will be installed
---> Package salt.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: systemd-units for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-units for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-python for package: salt-2015.8.3-1.el7.noarch
---> Package salt-master.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: systemd-python for package: salt-master-2015.8.3-1.el7.noarch
---> Package zeromq.x86_64 0:4.0.5-4.el7 will be installed
--> Running transaction check
---> Package python-requests.noarch 0:2.6.0-1.el7 will be installed
--> Processing Dependency: python-urllib3 >= 1.10.2-1 for package: python-requests-2.6.0-1.el7.noarch
--> Processing Dependency: python-chardet >= 2.2.1-1 for package: python-requests-2.6.0-1.el7.noarch
---> Package python26-backports.x86_64 0:1.0-3.14.amzn1 will be installed
---> Package python26-setuptools.noarch 0:12.2-1.30.amzn1 will be installed
---> Package salt.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: systemd-units for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-units for package: salt-2015.8.3-1.el7.noarch
--> Processing Dependency: systemd-python for package: salt-2015.8.3-1.el7.noarch
---> Package salt-master.noarch 0:2015.8.3-1.el7 will be installed
--> Processing Dependency: systemd-python for package: salt-master-2015.8.3-1.el7.noarch
--> Finished Dependency Resolution
Error: Package: python-requests-2.6.0-1.el7.noarch (saltstack-repo)
           Requires: python-chardet >= 2.2.1-1
           Available: python26-chardet-2.0.1-7.7.amzn1.noarch (amzn-main)
               python-chardet = 2.0.1-7.7.amzn1
Error: Package: salt-2015.8.3-1.el7.noarch (saltstack-repo)
           Requires: systemd-units
Error: Package: salt-master-2015.8.3-1.el7.noarch (saltstack-repo)
           Requires: systemd-python
Error: Package: python-requests-2.6.0-1.el7.noarch (saltstack-repo)
           Requires: python-urllib3 >= 1.10.2-1
           Available: python26-urllib3-1.8.2-1.5.amzn1.noarch (amzn-main)
               python-urllib3 = 1.8.2-1.5.amzn1
Error: Package: salt-2015.8.3-1.el7.noarch (saltstack-repo)
           Requires: systemd-python
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
````

**Comment by lorengordon** [comment]
Try using the new saltstack repo, https://repo.saltstack.com/#rhel

**Comment by edhgoose** [comment]
@lorengordon Those instructions are the same as the instructions I referenced, although they don't refer to Amazon any more.

However .... the bootstrap script referenced on that page (under Bootstrap - Multi Platform) does work! Thank you.

For anyone else who has this issue, I couldn't get the `install_salt.sh` script to run while I had the previous EPEL version of Salt installed. I had to run `sudo yum remove python-urllib3`, which removed the following:

- salt-master
- salt-minion
- python26-requests
- python26-boto
- salt

I ran the `sudo yum remove python-urllib3` because this was the error I got when trying to run the install_salt.sh script:

````
sudo sh install_salt.sh -P -M
 *  INFO: PIPED THROUGH install_salt.sh -- Version 2015.11.09

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   3.14.35-28.38.amzn1.x86_64
 *  INFO:   Distribution: Amazon Linux AMI 2015.09

 *  INFO: Installing minion
 *  INFO: Installing master
 *  INFO: Found function install_amazon_linux_ami_deps
 *  INFO: Found function install_amazon_linux_ami_stable
 *  INFO: Found function install_amazon_linux_ami_stable_post
 *  INFO: Found function install_amazon_linux_ami_restart_daemons
 *  INFO: Found function daemons_running
 *  INFO: Running install_amazon_linux_ami_deps()
Retrieving http://mirrors.kernel.org/fedora-epel/6/x86_64/epel-release-6-8.noarch.rpm
Preparing...                          ########################################
Updating / installing...
epel-release-6-8                      ########################################
Cleaning up / removing...
epel-release-6-8.9.amzn1              ########################################
 *  INFO: Adding SaltStack's COPR repository
Loaded plugins: priorities, update-motd, upgrade-helper
948 packages excluded due to repository priority protections
Package python26-m2crypto-0.21.1-15.14.amzn1.x86_64 already installed and latest version
Package python-crypto-2.6.1-1.el7.x86_64 is obsoleted by python26-crypto-2.6.1-1.12.amzn1.x86_64 which is already installed
Package python-ordereddict-1.1-2.el6.noarch is obsoleted by python26-2.6.9-2.84.amzn1.x86_64 which is already installed
Package python26-jinja2-2.7.2-2.15.amzn1.noarch already installed and latest version
Resolving Dependencies
--> Running transaction check
---> Package PyYAML.x86_64 0:3.11-1.el7 will be installed
---> Package python-msgpack.x86_64 0:0.4.6-1.el6 will be updated
---> Package python-msgpack.x86_64 0:0.4.6-1.el7 will be an update
---> Package python-requests.noarch 0:2.6.0-1.el7 will be installed
--> Processing Dependency: python-urllib3 >= 1.10.2-1 for package: python-requests-2.6.0-1.el7.noarch
--> Processing Dependency: python-chardet >= 2.2.1-1 for package: python-requests-2.6.0-1.el7.noarch
---> Package python-zmq.x86_64 0:14.3.1-1.el6 will be updated
---> Package python-zmq.x86_64 0:14.7.0-1.el7 will be an update
--> Processing Dependency: libzmq.so.4()(64bit) for package: python-zmq-14.7.0-1.el7.x86_64
--> Running transaction check
---> Package python-requests.noarch 0:2.6.0-1.el7 will be installed
--> Processing Dependency: python-urllib3 >= 1.10.2-1 for package: python-requests-2.6.0-1.el7.noarch
--> Processing Dependency: python-chardet >= 2.2.1-1 for package: python-requests-2.6.0-1.el7.noarch
---> Package zeromq.x86_64 0:4.0.5-4.el7 will be installed
--> Finished Dependency Resolution
Error: Package: python-requests-2.6.0-1.el7.noarch (saltstack-repo)
           Requires: python-chardet >= 2.2.1-1
           Installed: python26-chardet-2.0.1-7.7.amzn1.noarch (@amzn-main)
               python-chardet = 2.0.1-7.7.amzn1
Error: Package: python-requests-2.6.0-1.el7.noarch (saltstack-repo)
           Requires: python-urllib3 >= 1.10.2-1
           Installed: python26-urllib3-1.8.2-1.5.amzn1.noarch (@amzn-main)
               python-urllib3 = 1.8.2-1.5.amzn1
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
 * ERROR: Failed to run install_amazon_linux_ami_deps()!!!
````

Thanks again @lorengordon

**Comment by lorengordon** [comment]
Sorry about the duplicate instructions, I admittedly didn't click on the link you provided; I saw the url went to the docs site rather than the repo site, and I just assumed the docs site still had old instructions. :shamefaced: 

Glad you got it working though.

**Comment by edhgoose** [comment]
Sorry, I've realised I haven't got it working at all. I've actually realised I've gone backwards!

The command line version only installs this version:

````
$ salt --version
salt 2015.5.5 (Lithium)
````

And the second thing I've realised (or remembered rather) is that the `python26-boto` library is installed to make the ec2_tags plugin (in the contrib repo: https://github.com/saltstack/salt-contrib/tree/master/grains) work.

So ... back where I started:

1. The instructions are out of date
2. The RedHat instructions don't work
3. The EPEL instructions do work, but install an old version
4. The install by command line LOOKS like it works, but installs an even older version.

:(

**Comment by jtand** [comment]
@dmurphy18 @terminalmage

**Comment by tkunicki** [comment]
I ran across this last night.  I had to modify the instructions for use with Amazon Linux AMI 2015.9.1 to use the following:
```
[saltstack-repo]
name=SaltStack repo for RHEL/CentOS 6
baseurl=https://repo.saltstack.com/yum/redhat/6/$basearch/latest
enabled=1
gpgcheck=1
gpgkey=https://repo.saltstack.com/yum/redhat/6/$basearch/latest/SALTSTACK-GPG-KEY.pub
```

The current [documentation for RHEL et al.](https://docs.saltstack.com/en/latest/topics/installation/rhel.html) specifies the use of
```
[saltstack-repo]
name=SaltStack repo for RHEL/CentOS $releasever
baseurl=https://repo.saltstack.com/yum/redhat/$releasever/$basearch/latest
enabled=1
gpgcheck=1
gpgkey=https://repo.saltstack.com/yum/redhat/$releasever/$basearch/latest/SALTSTACK-GPG-KEY.pub
``` 

I think the issue relates to this:
```
[ec2-user@salt ~]$ cat /etc/yum.conf | grep releasever=
releasever=latest
```
The file `/etc/yum.conf`, I believe, was provided by the base AMU and `releasever` is specified as `latest` even though this linux version is based on version 6.  I am guessing 'latest` is currently aliased to version 7.

**Comment by edhgoose** [comment]
I can confirm that I also have been able to install Salt latest using 6 instead of $releasever. Thank you @tkunicki!

Is this a documentation bug (i.e. I should submit a PR to adjust [this installation page](https://github.com/saltstack/salt/blob/develop/doc/topics/installation/rhel.rst) to say "Version 6 or Amazon Linux", or is there an alternative fix that would be better?

**Comment by lorengordon** [comment]
Oh geez, ok, right. 

https://github.com/saltstack/salt/issues/27281#issuecomment-142080230
https://github.com/saltstack/salt/issues/28190#issuecomment-150214234
https://github.com/saltstack/salt/pull/29102#issuecomment-158778197

It would be nice to have redirects for the known $releasever values so this issue would go away.

**Comment by damon-atkins** [comment]
You should look at using yum mirror list option, which would allow you to use a cgi script to take inputs and then send back a url, it would be a better option than creating lots of sym links in your repo. And also allow you to collect data about what platforms are using yum.  i.e. mirrorlist=yum.cgi?releaserver=$releaserver;arch=$arch;basearch=$basearch
which would then return https://repo.saltstack.com/yum/redhat/.......

**Comment by lorengordon** [comment]
@damon-atkins, nice, yeah, that would be even better.

**Comment by brunsgaard** [comment]
On new EC2 instance (Amazon Linux AMI 2015.09.1 (HVM), SSD Volume Type - ami-bc5b48d0)

```
sudo yum update
sudo rpm --import https://repo.saltstack.com/yum/redhat/6/x86_64/latest/SALTSTACK-GPG-KEY.pub
sudo vim /etc/yum.repos.d/saltstack.repo
```
insert into `/etc/yum.repos.d/saltstack.repo`
```
[saltstack-repo]
name=SaltStack repo for RHEL/CentOS 6
baseurl=https://repo.saltstack.com/yum/redhat/6/$basearch/latest
enabled=1
gpgcheck=1
gpgkey=https://repo.saltstack.com/yum/redhat/6/$basearch/latest/SALTSTACK-GPG-KEY.pub
```
Proceed with the following commands 
```
sudo yum clean expire-cache
sudo yum update
yum install salt-minion
```

Now you are happy

**Comment by dmurphy18** [comment]
@edhgoose @brunsgaard @damon-atkins @lorengordon Just wanted to make you aware that yesterday, uploaded an install repo rpm for Amazon Linux and documentation on Amazon Linux (own tab) on repo.saltstack.com.  The install repo rpm installs the latest based on Redhat 6, but adjusts the baseurl similar to that stated above by bunsgaard.

Hopefully this will ease the installation of Salt onto Amazon Linux.  Work is ongoing to get the working develop bootstrap changes into a stable bootstrap release also.

With this install repo rpm and information, can this issue be closed ?

**Comment by lorengordon** [comment]
@dmurphy18, the instructions look pretty good to me. Only nitpick I have is that 'Pin to Minor Release' under Amazon Linux still identifies RHEL/CentOS in the 'name' field of the repo definition, instead of Amazon Linux.

i.e., this:
```
[saltstack-amzn-repo]
name=SaltStack repo for RHEL/CentOS $releasever
baseurl=https://repo.saltstack.com/yum/redhat/6/$basearch/archive/2015.8.7
enabled=1
gpgcheck=1
gpgkey=https://repo.saltstack.com/yum/redhat/6/$basearch/archive/2015.8.7/SALTSTACK-GPG-KEY.pub
```

should be:
```
[saltstack-amzn-repo]
name=SaltStack repo for Amazon Linux
baseurl=https://repo.saltstack.com/yum/redhat/6/$basearch/archive/2015.8.7
enabled=1
gpgcheck=1
gpgkey=https://repo.saltstack.com/yum/redhat/6/$basearch/archive/2015.8.7/SALTSTACK-GPG-KEY.pub
```

**Comment by dmurphy18** [comment]
@lorengordon Thanks for that catch, missed it when we proof-read, even after multiple sets of eyes. Will fix it.

@jacobhammons Can you up date the docs and I shall post the fix, see above for Amazon

**Comment by lorengordon** [comment]
@dmurphy18, cool. One more thing related to the instructions, maybe a tangent issue though. But if I were installing a release rpm that was versioned to the major release, e.g. `salt-amzn-repo-2015.8-1.ami.noarch.rpm`, then I'd expect that it would pin salt to that release version, rather than installing the latest salt version and upgrading salt across major releases. Maybe use a different versioning scheme for the release rpms?

**Comment by dmurphy18** [comment]
@lorengordon Yes, not the first to mention this but for now want to keep the version in to help in identifying  which install repo rpm, esp. if there is a number of them in the same directory, noting that 2015.8 shall not be the only one soon. However I shall revisit this with the next major branch release.

**Comment by dmurphy18** [comment]
@lorengordon Documentation fix is live now, I am going to close this. If there is a new  item, you see, can you open a new issue since I believe that the issue stated here has been addressed.
